jib-core/src/main/java/com/google/cloud/tools/jib/frontend/JavaLayerConfigurations.java

```diff
@@ -60,6 +64,40 @@ String getName() {
   /** Builds with each layer's files. */
   public static class Builder {
 
+    /**
+     * Adds files to a layer selectively and recursively. {@code sourceRoot} must be a directory.
+     * Empty directories will always be added regardless of {@code pathFilter}, except for {@code
+     * sourceRoot}. If {@code sourceRoot} is empty, nothing will be added.
+     *
+     * <p>The contents of {@code sourceRoot} will be placed into {@code basePathInContainer}. For
+     * example, if {@code sourceRoot} is {@code /usr/home}, {@code /usr/home/passwd} exists locally,
+     * and {@code basePathInContainer} is {@code /etc}, then the image will have {@code
+     * /etc/passwd}.
+     *
+     * @param sourceRoot root directory whose contents will be added
+     * @param pathFilter only the files satisfying the filter will be added, unless the files are
+     *     directories
+     * @param basePathInContainer directory in the layer into which the source contents are added
+     * @param layerBuilder the {@link LayerConfiguration.Builder}
+     * @throws IOException error while listing directories
+     * @throws NotDirectoryException if {@code sourceRoot} is not a directory
+     */
+    @VisibleForTesting
+    static void addFilesRoot(
```

Comment:
Looks like all of `JavaContainerBuilder`, `LayerConfiguration.Builder`, and `JavaLayerConfigurations.Builder` should be re-worked. I feel the current API is not very friendly for some of our use cases, where files in the same directory hierarchy should be separated out into different layers. Also, the lowest level `addEntry` and `addEntryRecursive` of `LayerConfiguration.Builder` is constrained in that it's very cumbersome to recursively copy files into the container root. (For that matter, you have to list files first and then call `addEntryRecursive` multiple times.)

Reply:
> I think how we could rework those APIs is to have `LayerConfiguration.Builder` only add the low-level sourceFile-extractionPath pairs,  and to have a higher-level helper

This makes sense. The low-level pair (with no recursion) is very limited but acceptable assuming the users won't work on the low-level `LayerConfiguration` directly. The helper can solve the recursive copy with filtering.

Now the question is, where this help should be placed by which name. I realized there is one annoyance with helper methods in another class: if such methods are `static`, they should accept a builder instance of  `LayerConfiguration.Builder` as an argument to work on it. This may be OK but feels a bit awkward personally, which makes me think maybe I should just put the helper methods under `LayerConfiguration.Builder` directly. If we don't have the helper methods as `static` (i.e., a stateful helper for building a `LayerConfiguration`), it becomes another `LayerConfiguration` builder that is more versatile than the low-level `LayerConfiguration.Builder`.

Also, currently `JavaLayerConfigurations` is another abstract layer on top of `LayerConfiguration`, which can also act as a helper, so we need to consider that too for deciding a suitable place for the helper class.